storage/local/storage.go

```diff
@@ -559,25 +573,29 @@ func (s *memorySeriesStorage) Append(sample *clientmodel.Sample) {
 	s.incNumChunksToPersist(completedChunksCount)
 }
 
-func (s *memorySeriesStorage) getOrCreateSeries(fp clientmodel.Fingerprint, m clientmodel.Metric) *memorySeries {
+func (s *memorySeriesStorage) getOrCreateSeries(fp clientmodel.Fingerprint, m clientmodel.Metric) (series *memorySeries, lastAppendTime clientmodel.Timestamp) {
 	series, ok := s.fpToSeries.get(fp)
 	if !ok {
-		unarchived, firstTime, err := s.persistence.unarchiveMetric(fp)
+		unarchived, firstTime, lastTime, err := s.persistence.unarchiveMetric(fp)
 		if err != nil {
 			log.Errorf("Error unarchiving fingerprint %v: %v", fp, err)
 		}
 		if unarchived {
 			s.seriesOps.WithLabelValues(unarchive).Inc()
+			lastAppendTime = lastTime
 		} else {
 			// This was a genuinely new series, so index the metric.
 			s.persistence.indexMetric(fp, m)
 			s.seriesOps.WithLabelValues(create).Inc()
+			lastAppendTime = clientmodel.Earliest
 		}
 		series = newMemorySeries(m, !unarchived, firstTime)
 		s.fpToSeries.put(fp, series)
 		s.numSeries.Inc()
+	} else {
+		lastAppendTime = series.head().lastTime()
```

Comment:
This is a fairly expensive operation. It involves locking the timestamp and creation of a chunk iterator.
And this is the hottest spot in the code, i.e. almost all appends will go through this.
We need to remember the last append time in the series object (and probably want to include it in the checkpoint, too).